docs/proposals/admission_control_changes.md

```diff
@@ -0,0 +1,485 @@
+   setup in a structured configuration, once we've fully moved over)
+
+You would specify `--admission-plugin` once per plugin that you want to
+configure. An example of this would be
+
+```
+--admission-plugin AlwaysPullImages=/path/to/config.yaml --admission-plugin MyPlugin=/some/other/config.yaml
+```
+
+An example of a structured configuration for admission control inside an
+apiserver configuration file might look like this:
+
+```yaml
+admissionConfig:
+  AlwaysPullImages:
+    configurationFile: /path/to/config.yaml
+  MyPlugin:
+    configurationFile: /some/other/config.yaml
+  
+```
+
+## Ability to enable/disable individual plugins
+
+With a fixed list of plugins enabled by default, administrators will need to be
+able to override the defaults and disable certain plugins that are normally on,
+and enable plugins that are normally off. We are proposing that the admission
+control section in the apiserver configuration file include support for this. We
+can use `enabled` as seen below:
+
+```yaml
+admissionConfig:
+  AlwaysPullImages:
+    # disabled by default, so enable it
+    enabled: true
+    configurationFile: /path/to/config.yaml
+  MyPlugin:
+    # enabled by default, so disable it
+    enabled: false
+    configurationFile: /some/other/config.yaml
+```
+
+Note, the enabling/disabling mechanism described in this proposal only works
+when using a structured configuration setup for the apiserver. The
+`--admission-plugin` flag described above does not provide a means to enable or
+disable a plugin (TODO: should it????).
+
+## Code changes
+
+### Admission interfaces
+
+The current interface in `pkg/admission/interfaces.go` looks like this:
+
+```go
+type Interface interface {
+  Admit(a Attributes) (err error)
+  Handles(operation Operation) bool
+}
+```
+
+We'd like to break this into multiple interfaces, one per phase, called at the appropriate times
+within the request lifecycle.
+
+```go
+type Defaulter interface {
+  // ApplyDefaults is executed after the request body is deserialized into an
+  // object for create requests (TODO do we want this for update/patch too?).
+  // Plugins are free to modify the object as they see fit. If an error is
+  // returned, the request is marked as failed and no changes are persisted.
+  ApplyDefaults(a Attributes) error
+}
+
+type Validator interface {
+  // Validate is executed after the appropriate `rest.Storage` handler's
+  // validation occurs. This gives the plugin the opportunity to validate the
+  // object after the handler has validated it, but before it is persisted. If
+  // an error is returned, the request is marked as failed and no changes are
+  // persisted.
+  Validate(a Attributes) error
+}
+
+
+type ConstraintEnforcer interface {
+  // EnforceConstraints is executed after all Validator plugins have executed.
+  // This gives plugins a chance to decide if the request should be allowed to
+  // proceed after the object has passed `rest.Storage` handler validation.
```

Comment:
This is where we'd do quota, for example. If an object isn't valid, we want it to fail either in the registry package's validation, or in admission validation, *before* we ever get to checking quota.